File: DASHEmbed/Framework/DASHViewController.m
//
//  DASHViewController.m
//  DASHEmbed
//
//  Created by Ryan Gant on 11/24/19.
//  Copyright © 2019 DASH. All rights reserved.
//

#import "DASHViewController.h"
#import "DASHConfig.h"
#import "DASHUserInfo.h"
#import "NSBundle+DASHAdditions.h"
@import WebKit;

static NSString *const DASBaseDevelopmentURLString = @"https://dev-web.dashapp.io/app";
static NSString *const DASBaseURLString = @"https://web.dashapp.io/app";
static NSString *const DASApplicationQueryName = @"appId";
static NSString *const DASPlatformQueryName = @"platformId";
static NSString *const DASPlatformQueryValue = @"ios";
static NSString *const DASEmailQueryName = @"email";
static NSString *const DASPushQueryName = @"pushId";

@interface DASHViewController () <WKNavigationDelegate, WKUIDelegate>

@property (nonatomic, strong) WKWebView *webView;
@property (nonatomic, strong) DASHConfig *config;
@property (nonatomic, strong) DASHUserInfo *userInfo;
@property (nonatomic, strong) NSDictionary<NSString *, id> *notificationData;

@end

@implementation DASHViewController

+ (instancetype)instantiateWithConfig:(DASHConfig *)config userInfo:(DASHUserInfo *)userInfo notificationData:(NSDictionary<NSString *, id> *)notificationData {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"DASH" bundle:[NSBundle frameworkResourceBundle]];
    DASHViewController *viewController = [storyboard instantiateInitialViewController];
    
    viewController.config = config;
    viewController.userInfo = userInfo;
    viewController.notificationData = notificationData;
    
    return viewController;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupWebView];
    [self loadWebView];
}

#pragma mark - Public

- (void)updateNotificationData:(NSDictionary<NSString *, id> *)notificationData {
    _notificationData = notificationData;
    [self reloadInterfaceStartFromBeginning:YES];
}

- (void)reloadInterfaceStartFromBeginning:(BOOL)startFromBeginning {
    if (startFromBeginning) {
        [self loadWebView];
    } else {
        [self.webView reload];
    }
}

#pragma mark - Private

- (void)setupWebView {
    WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
    self.webView = [[WKWebView alloc] initWithFrame:self.view.bounds configuration:configuration];
    self.webView.navigationDelegate = self;
    self.webView.UIDelegate = self;
    self.webView.translatesAutoresizingMaskIntoConstraints = false;
    [self.view addSubview:self.webView];
    NSDictionary<NSString *, UIView *> *views = @{@"webView": self.webView};
    NSArray<NSLayoutConstraint *> *horizontalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[webView]|" options:0 metrics:nil views:views];
    NSArray<NSLayoutConstraint *> *verticalConstraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[webView]|" options:0 metrics:nil views:views];
    [self.view addConstraints:horizontalConstraints];
    [self.view addConstraints:verticalConstraints];
}

- (void)loadWebView {
    if (!self.config) {
        NSLog(@"A DASHConfig is required before using DASHViewController");
        NSError *unableToLoadError = [NSError errorWithDomain:@"io.dashapp.dash" code:1 userInfo:@{NSLocalizedFailureReasonErrorKey: @"Unable to load DASH. Please try again later."}];
        [self.delegate dashViewController:self didFailWithError:unableToLoadError];
        return;
    }
    
    NSString *urlString = self.config.useDevelopmentServers ? DASBaseDevelopmentURLString : DASBaseURLString;
    NSURL *url = [NSURL URLWithString:urlString];
    if (url) {
        NSURLComponents *urlComponents = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
        
        NSMutableArray<NSURLQueryItem *> *queryItems = [NSMutableArray array];
        
        //Add default query items
        NSURLQueryItem *applicationQueryItem = [NSURLQueryItem queryItemWithName:DASApplicationQueryName value:self.config.appId];
        [queryItems addObject:applicationQueryItem];
        NSURLQueryItem *platformQueryItem = [NSURLQueryItem queryItemWithName:DASPlatformQueryName value:DASPlatformQueryValue];
        [queryItems addObject:platformQueryItem];
        
        //Add user email if provided
        if (self.userInfo.userEmail) {
            NSURLQueryItem *userEmailQueryItem = [NSURLQueryItem queryItemWithName:DASEmailQueryName value:self.userInfo.userEmail];
            [queryItems addObject:userEmailQueryItem];
        }
        
        //Add push token if provided
        if (self.userInfo.pushTokenString) {
            NSURLQueryItem *pushQueryItem = [NSURLQueryItem queryItemWithName:DASPushQueryName value:self.userInfo.pushTokenString];
            [queryItems addObject:pushQueryItem];
        }
        
        //Add notification data as query parameters if present
        if (self.notificationData) {
            [self.notificationData enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, id  _Nonnull obj, BOOL * _Nonnull stop) {
                NSURLQueryItem *notificationQuery = [NSURLQueryItem queryItemWithName:key value:[NSString stringWithFormat:@"%@", obj]];
                [queryItems addObject:notificationQuery];
            }];
        }
        
        urlComponents.queryItems = queryItems;
        NSURL *fullUrl = urlComponents.URL;
        if (fullUrl) {
            NSURLRequest *request = [NSURLRequest requestWithURL:fullUrl];
            [self.webView loadRequest:request];
        }
    }
}

#pragma mark - WKNavigationDelegate

- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [self.delegate dashViewController:self didFailWithError:error];
}

#pragma mark - WKUIDelegate

- (WKWebView *)webView:(WKWebView *)webView createWebViewWithConfiguration:(WKWebViewConfiguration *)configuration forNavigationAction:(WKNavigationAction *)navigationAction windowFeatures:(WKWindowFeatures *)windowFeatures {
    NSURL *url = navigationAction.request.URL;
    if ([[UIApplication sharedApplication] canOpenURL:url]) {
        [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
    }
    
    return nil;
}

@end

File: DASHEmbed/Framework/DASHViewController.swift
//
//  ViewController.swift
//  DASHEmbed
//
//  Copyright © 2018 DASH. All rights reserved.
//

import UIKit
import WebKit

public protocol DASHViewControllerDelegate: class {
    /// Called when the DASH view controller encounters an error.
    func dashViewController(_ dashViewController: DASHViewController, didFailWith error: DASH.Error)
}

public class DASHViewController: UIViewController {
    
    private let baseURLString = "https://dev-web.dashapp.io/app"
    private let applicationQueryName = "appId"
    private let platformQueryName = "platformId"
    private let platformQueryValue = "ios"
    private let emailQueryName = "email"
    private let pushQueryName = "pushId"

    private var webView: WKWebView!
    private var config: DASHConfig?
    private var userInfo: DASH.UserInfo?
    private var notificationData: [String: Any]?
    
    public weak var delegate: DASHViewControllerDelegate?
    
    static func instantiate(with config: DASHConfig, userInfo: DASH.UserInfo, notificationData: [String: Any]?) -> DASHViewController? {
        let storyboard = UIStoryboard(name: "DASH", bundle: Bundle.frameworkResourceBundle)
        guard let viewController = storyboard.instantiateInitialViewController() as? DASHViewController else {
            print("Unable to instantiate DASHViewController.")
            return nil
        }
        
        viewController.config = config
        viewController.userInfo = userInfo
        viewController.notificationData = notificationData
        
        return viewController
    }

    override public func viewDidLoad() {
        super.viewDidLoad()
        setupWebView()
        loadWebView()
    }
    
    /// Updates the view controller with notification data. Used internally.
    public func updateNotificationData(with data: [String: Any]?) {
        notificationData = data
        reloadInterface(startFromBeginning: true) //Reload the web view to the correct page
    }
    
    /// Refreshes the current page. If startFromBeginning is true, the interface is reloaded to the beginning state.
    public func reloadInterface(startFromBeginning: Bool = false) {
        if startFromBeginning {
            loadWebView()
        } else {
            webView.reload()
        }
    }
    
    // MARK: Private
    
    private func setupWebView() {
        // Create WKWebView in code, because IB has issues before iOS 11
        let configuration = WKWebViewConfiguration()
        webView = WKWebView(frame: view.bounds, configuration: configuration)
        webView.navigationDelegate = self
        webView.translatesAutoresizingMaskIntoConstraints = false
        
        view.addSubview(webView)
        let horizontalContraints = NSLayoutConstraint.constraints(withVisualFormat: "H:|[webView]|",
                                                                  options: [],
                                                                  metrics: nil,
                                                                  views: ["webView": webView])
        let verticalConstraints = NSLayoutConstraint.constraints(withVisualFormat: "V:|[webView]|",
                                                                 options: [],
                                                                 metrics: nil,
                                                                 views: ["webView": webView])
        view.addConstraints(horizontalContraints)
        view.addConstraints(verticalConstraints)
    }
    
    private func loadWebView() {
        guard let config = config else {
            print("A DASHConfig is required before using DASHViewController")
            delegate?.dashViewController(self, didFailWith: .unableToLoad)
            return
        }
        
        if let url = URL(string: baseURLString) {
            var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false)
            
            var queryItems = [URLQueryItem]()
            
            //Add default query items
            let applicationQueryItem = URLQueryItem(name: applicationQueryName, value: config.appId)
            queryItems.append(applicationQueryItem)
            let platformQueryItem = URLQueryItem(name: platformQueryName, value: platformQueryValue)
            queryItems.append(platformQueryItem)
            
            //Add user email if provided
            if let userEmail = userInfo?.userEmail {
                let userEmailQueryItem = URLQueryItem(name: emailQueryName, value: userEmail)
                queryItems.append(userEmailQueryItem)
            }
            
            //Add push token if provided
            if let pushToken = userInfo?.pushTokenString {
                let pushQueryItem = URLQueryItem(name: pushQueryName, value: pushToken)
                queryItems.append(pushQueryItem)
            }
            
            //Add the notification data as query parameters if present
            if let notificationData = notificationData {
                for (key, value) in notificationData {
                    let notificationQuery = URLQueryItem(name: key, value: String(describing: value))
                    queryItems.append(notificationQuery)
                }
            }
            
            urlComponents?.queryItems = queryItems
            
            if let fullURL = urlComponents?.url {
                let urlRequest = URLRequest(url: fullURL)
                webView.load(urlRequest)
            }
        }
    }
}

extension DASHViewController: WKNavigationDelegate {
    public func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        let nsError = error as NSError
        if nsError.code == NSURLErrorNotConnectedToInternet {
            delegate?.dashViewController(self, didFailWith: .noInternet)
            return
        }
        
        delegate?.dashViewController(self, didFailWith: .unableToLoad)
    }
}
